### ISSUE TITLE
Remapping key to space

### ISSUE THREAD

**Issue by GabenCrying** [autokey/autokey#702]
### Has this issue already been reported?

- [X] I have searched through the existing issues.

### Is this a question rather than an issue?

- [X] This is not a question.

### What type of issue is this?

UI/Usability

### Which Linux distribution did you use?

Linux arch 5.18.3-arch1-1

### Which AutoKey GUI did you use?

Qt

### Which AutoKey version did you use?

autokey-qt Version 0.96.0

### How did you install AutoKey?

AUR

### Can you briefly describe the issue?

Cannot bind a key to space bar 

I'm trying remap a key to press space, can't find ANYWHERE how to do this

### Can the issue be reproduced?

Always

### What are the steps to reproduce the issue?

1. Just pressing SPACE in the phase says it's empty
2. Typing SPACE inserts the word SPACE when pressing the key
3.  (space character) inserts (space character) when pressing the key
4.  https://github.com/autokey/autokey/wiki/Special-Keys as such

### What should have happened?

Press desired key presses space

### What actually happened?

Inserts anything I type in the phrase box

### Anything else?

Sorry if I'm not clear, basically, I am trying to press np_end and have it activate the space bar. I am playing a bad game with no control settings, lol. I'm a numpad gamer.

**Comment by josephj11** [comment]
I believe I have recreated this. The problem looks as if AutoKey is trimming the phrase content so that specifying a phrase whose text is exclusively one or more spaces is not possible.

A developer will have to look at this to see what can be done to make this possible.

As a workaround, you can write a script instead of a phrase. All it will need is:

    keyboard.send_keys(" ")

**Comment by josephj11** [comment]
See: phrasepage.py lines 101 - 104 which calls `str(self.phraseText.toPlainText())`.

This appears to be a Qt built-in method.

The phrase content would have to be examined before this call. However, I have no idea what allowing an all blank phrase would do in the rest of the program and there is the possibility that some complex phrase with macros, etc. could generate all blank text, so just looking for all spaces wouldn't be sufficient.

**Comment by Elliria** [comment]
Interesting. I just tried it and it wouldn't let me create a phrase with just a space in it, either.

**Work-around:**

A temporary work-around would be to use an [AutoKey special key](https://github.com/autokey/autokey/wiki/Special-Keys) in the phrase so that the AutoKey API will insert the space for you.That page recommends typing a space into some angled brackets to create a space, but that doesn't work in this case and will print the brackets and the space.

Luckily, if you use a keycode in the brackets instead, that works, so a phrase with just `<code65>` in it should insert a space.

In the event that your keyboard is mapped differently from mine and `<code65>` inserts something other than a space, you'll need to find out what your space bar's keycode is and use it instead. The [Customized special keys](https://github.com/autokey/autokey/wiki/Special-Keys#customized-special-keys) section of the above page demonstrates a few ways to get a keycode.

**Comment by ineuw** [comment]
Is there a possibility of an actual example?  Because, space in itself can be a hotkey.

**Comment by Elliria** [comment]
It occurs to me that a phrase might not work with a game. If it doesn't work, Joe's solution of creating a script that contains this should work:
```
keyboard.send_key(" ")
```
If that also doesn't work, you could try creating a script that contains this:
```
keyboard.press_key(" ")
keyboard.release_key(" ")
```

**Comment by ineuw** [comment]
@Elliria, in my gtk world I never had any luck with ```keyboard.send_key(" ")``` but must always use ```keyboard.sendkeys(" ")``` in any 95 or 96 versions.

**Comment by Elliria** [comment]
Whoops - my apologies. That's exactly what Joe recommended. Thanks for catching it. It's now edited.

**Comment by ineuw** [comment]
@Elliria please don't apologize. :-) I would like to see the OP's  issue in a fictitious example. To be honest, I can't understand it.

**Comment by Elliria** [comment]
Ah, I can provide that since my son also likes to play games with the numpad. The OP wants the **end** key to act as a space bar in a video game.

**Comment by josephj11** [comment]
@Elliria Your workaround is quite original. Did you actually test it though? It might work the way bash does and see the reinterpreted text as all blank and do something unexpected. I'm not clear on what it will do.

**Comment by josephj11** [comment]
Given the way the phrase definition code is written, this will most likely apply to other white space characters as well like Tab and, possibly, Enter. I have not tested it.

**Comment by ineuw** [comment]
Perhaps replacing the space with the Unicode character code, "\u0020"  will work better

**Comment by Elliria** [comment]
Ah, yes. I had forgotten about those Unicoode key-combinations, which would also work here. Those can be done in a couple of ways. In this one, you can generate a space with this key-combination followed by a space:
```
<ctrl>+<shift>+u+0020 
```
In this one, you can generate a space with the same key-combination followed by a plus sign:
```
<ctrl>+<shift>+u+0020+
```
They both do the same thing. It's just that the space is easy to overlook when debugging your code and easy to forget when pasting in the key-combination. I've messed up in both of those ways from time to time, so I prefer to use the plus sign.

**Comment by Elliria** [comment]
I've been testing all of these, @josephj11, and have been able to get AutoKey to fall over with just a space in a phrase and to print out something other than a space with various attempts and to successfully generate a space with each of the techniques I've been putting in here. I also tested out ineuw's suggestion to use Unicode and that successfully generates a space as well.

As you suspected, AutoKey won't let you create a phrase that contains just a tab or press of the Enter key, either. 
Those don't require keycodes or Unicode key-combinations, though, and can use AutoKey's special keys instead. A phrase that contains `<enter>` or `<tab>` will generate either of those.

_In case you'd like to geek out with me, I found something interesting while experimenting with these. Normally, any Unicode character I've managed to generate with a key-combination has required either a space or a plus sign at the end of the key-combination. If you were to generate a press of the Enter key with Unicode, however, you would definitely **not** put a space or plus sign at the end of the key-combination because AutoKey will insert a space there if you do, so you would use `<ctrl>+<shift>+u+000D` as is. The same is true whether you use it in a phrase or a script. This is the first time I've seen that sort of behavior and it might be good to keep in mind for any "misbehaving" Unicode characters in the future._

**Comment by josephj11** [comment]
@Elliria Thanks for the detailed followup. If you're feeling ambitious, this would make a nice addition to the wiki under the Unicode section. Probably just the examples would be enough.

Inspired weirdness:

I wondered what would happen if I attempted to emit a NUL character
   ` <ctrl>+<shift>+u+0020<ctrl>+<shift>+u+0000`

I got
    `+0020+0000+`

which was unexpected. Just another edge case that no one would probably do on purpose.

BTW, I don't know if they're in Unicode, but ASCII also contains vertical tab VT (0x0B), formfeed FF (0x0C), and carriage return CR (0x0D) which are also white space although they're pretty unlikely to be used on modern Linux systems for anything in a GUI (although Linux terminals do honor FF as a keyboard shortcut to clear the screen and if you print one, most printers will eject the current page or a new one if there is no current page and CR works the same as a newline. VT doesn't seem to do anything, at least in konsole). However, other programs including AutoKey can use them for keyboard shortcuts.

**Comment by Elliria** [comment]
Interesting. That one partly fails because you left off the plus or space at the end of each of the key-combinations (this is a perfect example of why I prefer having the plus be part of the syntax so that it's a visible reminder of its presence).

It can be written like this:
```
<ctrl>+<shift>+u+0020 <ctrl>+<shift>+u+0000 
```
Or like this:
```
<ctrl>+<shift>+u+0020+<ctrl>+<shift>+u+0000+
```
That doesn't fully solve it, though. It prints a space for the first one, but nothing for the second one when I believe it should print a space followed by a zero (at least that's what it prints if you print it in the Python interpreter).

**Comment by ineuw** [comment]
Try without the +. Haven't tried it currently, but I remember that it was a different result than "<ctrl>+<shift>"
```
keyboard.send_keys(<ctrl><shift>+"\u0020") 
```

**Comment by josephj11** [comment]
@Elliria Oh, I misunderstood that. I thought you only needed the blank or + after the last one. I don't know what the Python interpreter is thinking, but a NUL is definitely not a zero. It is one of very few characters not allowed in Linux file names and is the character the C programming language uses to denote the end of a variable length character string. That gets used a lot in Linux, so confusing the two could lead to lots of problems that would be hard to debug.

**Comment by josephj11** [comment]
A basic programming technique is the use of a "sentinel". It greatly simplifies writing loops over unknown input. It reduces the need for all sorts of character counting code and special cases. You write an indefinite loop that continues until it it encounters the sentinel (usually a character). For that to work, you need to be able to guarantee that the sentinel will never naturally occur in the data. That coupled with the common practice of providing new unused storage to a program (OS memory allocation) as a a block of data filled with zeros makes 0x00 an ideal choice for a sentinel. This also helps if something fails or otherwise returns no output by allowing the receiver to deal with it gracefully.

**Comment by Elliria** [comment]
Interesting. You've sent me down yet another deep rabbit-hole with all sorts of little nooks and crannies to explore. I can see where a NUL might be very useful as a sentinel. I've done some sentinel-like stuff in the past in various creative or "sideways" ways, but didn't realize they had a name or were a thing that can be done in a specific way. Apparently, there's a PEP for giving us a function that creates sentinel objects to make coding them simpler or more elegant.

None of the coding I've done has involved managing (or even taught me how to manage) memory, but from what you described, it sounds like I'm successfully avoiding something that's fraught with peril.

**Comment by josephj11** [comment]
I'm Late! I'm late for a very important date! LOL

Memory management is generally explicitly used in compiled languages like C. Interpreters such as Python usually do it behind the scenes for you and you may see references to things like heaps or garbage collection.

Yes, languages like C allow you to access anything anywhere (even if you do it by accident!) Debugging such things can be quite challenging. Part of it is handled by the memory hardware/firmware so that if you attempt to reference memory that's not allocated to your program you get a segment fault which usually terminates your program.

**Comment by ineuw** [comment]
Please correct me if I am wrong, but the Null character and the first 32 ASCII/ANSI characters existed since the 1920's. It was accepted by the IEEE with a strong push by IBM because of their emerging teletype system.

**Comment by josephj11** [comment]
@ineuw Wow! I can't imagine that ASCII is that old (possibly named that later). I'll have to go to Wikipedia and take a look. I know about them from actively using some of them in the `60s and `70s on my trusty ASR-33 teletype (with a paper tape reader/punch attachment) and doing some low level serial device communication stuff.

**Comment by ineuw** [comment]
I was even more surprised  . . . I am always curious when and what was adopted into wide use by society. I had to pick a hobby while I was reading and editing on English Wikisource. First, I wanted to know, when, and by whom, the concept of empirical evidence became the only acceptable method of proof in science. And, when I satisfied that curiosity, I followed the evolution and transformation of printed British English to American English.

**Comment by Elliria** [comment]
All but one of the languages I've learned don't have me managing memory, @josephj11, but that may be something to learn in the future. I'm still curious why these (apparently very old) characters (interesting stuff, @ineuw) are treated differently from other characters.

**Comment by josephj11** [comment]
In days of old. there was basically serial communications and that was it. Parallel ports came later, and much later, block devices like disk drives which use things like DMA - direct memory access. Input was restricted paper cards (80 or 96 column) with a bunch of rows for the bits I think there were 12 rows, so between 6 and 8 rows were used for character data and then there was a row for shift and for Ctrl and paper tapes that were 8 columns wide.. For quite a while even the eighth row was only used as a parity bit because serial line noise was a serious problem. When it stopped being a parity bit, it unlocked another 128 character codes which were generally regional in nature, so not universally reliable. (That was a mess!)

TL;DR: when all you've got is a string of bytes, you need some way to do basic functions like carriage return, linefeed, vertical tab and backspace. There's nowhere else they could come from unless they were defined within the character set itself. Similarly, there were needs for ways to delimit fields, records, and files as well as to handle data flow control. This was all done in those few 32 lower characters especially because discrete logic circuits (no chips yet!) could easily  be triggered  on and off by those bits. The exception to this was the RUBout key, later called DEL, which had all 7 bits on. It was used primarily with paper tapes to rub out (erase) single characters so you could correct a typo without having to start over with a whole new tape - which could be a serious problem. I personally knew of one huge paper tape that took 17 hours just to read! You did not want repunch that whole thing for one typo or if just one section became outdated! ASCII handled all of this in a pretty spectacular way for something so simple.
